File: dairytrack-web/src/admin/pages/produktivitas/MilkProductionLogs/Modal.js
import React, { useEffect, useCallback } from "react";
import {
  checkRawMilkExpired,
  getRawMilksByCowId,
} from "../../../../api/produktivitas/rawMilk";

const Modal = ({
  modalType,
  formData,
  setFormData,
  cows,
  handleSubmit,
  handleDelete,
  setModalType,
  selectedRawMilk,
  isProcessing,
  handleCowChange,
}) => {
  const fetchPreviousVolume = useCallback(
    async (cowId) => {
      if (!cowId) return;

      try {
        const rawMilks = await getRawMilksByCowId(cowId);
        const lastMilkData = rawMilks?.[0];
        setFormData((prev) => ({
          ...prev,
          previous_volume: lastMilkData?.volume_liters || 0,
        }));
      } catch (error) {
        console.error("Error fetching previous volume:", error.message);
        setFormData((prev) => ({
          ...prev,
          previous_volume: 0,
        }));
      }
    },
    [setFormData]
  );

  useEffect(() => {
    const timer = setTimeout(() => {
      if (formData.cow_id) {
        fetchPreviousVolume(formData.cow_id);
      }
    }, 500);

    return () => clearTimeout(timer);
  }, [formData.cow_id, fetchPreviousVolume]);

  const resetForm = useCallback(() => {
    setFormData({
      cow_id: "",
      production_time: "",
      volume_liters: "",
      previous_volume: 0,
      status: "fresh",
      lactation_status: false,
      lactation_phase: "Dry",
    });
  }, [setFormData]);

  return (
    <div
      className="modal fade show d-block"
      style={{
        background: isProcessing ? "rgba(0,0,0,0.8)" : "rgba(0,0,0,0.5)",
      }}
      tabIndex="-1"
      role="dialog"
      onClick={() => !isProcessing && setModalType(null)}
    >
      <div className="modal-dialog" onClick={(e) => e.stopPropagation()}>
        <div className="modal-content">
          <div className="modal-header">
            <h5 className="modal-title">
              {modalType === "create"
                ? "Add Raw Milk"
                : modalType === "edit"
                ? "Edit Raw Milk"
                : "Delete Confirmation"}
            </h5>
            <button
              type="button"
              className="btn-close"
              onClick={() => {
                setModalType(null);
                resetForm();
              }}
              disabled={isProcessing}
              aria-label="Close"
            ></button>
          </div>
          <div className="modal-body">
            {isProcessing ? (
              <div className="text-center">
                <div className="spinner-border text-primary" role="status">
                  <span className="visually-hidden">Loading...</span>
                </div>
                <p>Processing...</p>
              </div>
            ) : modalType === "delete" ? (
              <p>
                Are you sure you want to delete raw milk record{" "}
                <strong>{selectedRawMilk?.cow?.name || "this"}</strong>? This
                action cannot be undone.
              </p>
            ) : (
              <form
                onSubmit={(e) => {
                  e.preventDefault();
                  handleSubmit();
                }}
              >
                <div className="mb-3">
                  <label htmlFor="cow-select" className="form-label">
                    Cow
                  </label>
                  <select
                    id="cow-select"
                    className="form-control"
                    value={formData.cow_id}
                    onChange={(e) => handleCowChange(e.target.value)}
                    disabled={isProcessing || modalType === "edit"} // Nonaktifkan saat edit
                    required
                  >
                    <option value="">Select Cow</option>
                    {cows.map((cow) => (
                      <option key={cow.id} value={cow.id}>
                        {cow.name}
                      </option>
                    ))}
                  </select>
                </div>
                <div className="mb-3">
                  <label htmlFor="production-time" className="form-label">
                    Production Time
                  </label>
                  <input
                    id="production-time"
                    type="datetime-local"
                    className="form-control"
                    value={formData.production_time}
                    onChange={(e) =>
                      setFormData({
                        ...formData,
                        production_time: e.target.value,
                      })
                    }
                    disabled={isProcessing || modalType === "edit"} // Nonaktifkan saat edit
                    required
                  />
                </div>
                <div className="mb-3">
                  <label htmlFor="volume-liters" className="form-label">
                    Volume (Liters)
                  </label>
                  <input
                    id="volume-liters"
                    type="number"
                    className="form-control"
                    value={formData.volume_liters}
                    onChange={(e) =>
                      setFormData({
                        ...formData,
                        volume_liters: e.target.value,
                      })
                    }
                    disabled={isProcessing} // Tetap dapat diedit
                    min="0"
                    step="0.01"
                    required
                  />
                </div>
                <div className="mb-3">
                  <label htmlFor="previous-volume" className="form-label">
                    Previous Volume
                  </label>
                  <input
                    id="previous-volume"
                    type="number"
                    className="form-control"
                    value={formData.previous_volume || 0}
                    readOnly
                    disabled={true} // Selalu nonaktif
                  />
                </div>

                <div className="mb-3">
                  <label htmlFor="lactation-phase" className="form-label">
                    Lactation Phase
                  </label>
                  <select
                    id="lactation-phase"
                    className="form-control"
                    value={formData.lactation_phase}
                    onChange={(e) =>
                      setFormData({
                        ...formData,
                        lactation_phase: e.target.value,
                      })
                    }
                    disabled={!formData.lactation_status || isProcessing} // Disabled when lactation_status is unchecked
                  >
                    {!formData.lactation_status && (
                      <option value="Dry">Dry</option>
                    )}{" "}
                    {/* Show Dry only if lactation_status is unchecked */}
                    {formData.lactation_status && (
                      <>
                        <option value="Early">Early</option>
                        <option value="Mid">Mid</option>
                        <option value="Late">Late</option>
                      </>
                    )}
                  </select>
                </div>
                <div className="mb-3 d-flex align-items-center">
                  <div className="form-check">
                    <input
                      id="lactation-status"
                      type="checkbox"
                      className="form-check-input me-2"
                      checked={formData.lactation_status}
                      onChange={(e) => {
                        const isChecked = e.target.checked;
                        setFormData({
                          ...formData,
                          lactation_status: isChecked,
                          lactation_phase: isChecked ? "Early" : "Dry", // Reset lactation_phase based on checkbox
                        });
                      }}
                      disabled={isProcessing} // Tetap dapat diedit
                    />
                    <label
                      htmlFor="lactation-status"
                      className="form-check-label"
                    >
                      Lactation Status
                    </label>
                  </div>
                </div>
              </form>
            )}
          </div>
          <div className="modal-footer">
            <button
              type="button"
              className="btn btn-secondary"
              onClick={() => {
                setModalType(null);
                resetForm();
              }}
              disabled={isProcessing}
            >
              Cancel
            </button>
            {modalType === "delete" ? (
              <button
                type="button"
                className="btn btn-danger"
                onClick={handleDelete}
                disabled={isProcessing}
              >
                Delete
              </button>
            ) : (
              <button
                type="button"
                className="btn btn-primary"
                onClick={handleSubmit}
                disabled={isProcessing}
              >
                Save
              </button>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Modal;

File: dairytrack-mobile/lib/modules/produksi_susu/modal.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:dairy_track/config/api/produktivitas/rawMilk.dart';

class Modal extends StatefulWidget {
  final String? modalType;
  final Map<String, dynamic> formData;
  final Function(Map<String, dynamic>) setFormData;
  final List<dynamic> cows;
  final VoidCallback handleSubmit;
  final VoidCallback handleDelete;
  final Function(String?) setModalType;
  final dynamic selectedRawMilk;
  final bool isProcessing;
  final Function(String) handleCowChange;

  const Modal({
    Key? key,
    required this.modalType,
    required this.formData,
    required this.setFormData,
    required this.cows,
    required this.handleSubmit,
    required this.handleDelete,
    required this.setModalType,
    required this.selectedRawMilk,
    required this.isProcessing,
    required this.handleCowChange,
  }) : super(key: key);

  @override
  _ModalState createState() => _ModalState();
}

class _ModalState extends State<Modal> {
  final _formKey = GlobalKey<FormState>();
  late TextEditingController _productionTimeController;

  @override
  void initState() {
    super.initState();
    _productionTimeController = TextEditingController(
      text: widget.formData['production_time']?.isNotEmpty == true
          ? DateFormat('yyyy-MM-dd HH:mm')
              .format(DateTime.parse(widget.formData['production_time']))
          : '',
    );
  }

  @override
  void dispose() {
    _productionTimeController.dispose();
    super.dispose();
  }

  Future<void> _selectDateTime(BuildContext context) async {
    final DateTime? pickedDate = await showDatePicker(
      context: context,
      initialDate: DateTime.now(),
      firstDate: DateTime(2000),
      lastDate: DateTime(2100),
    );

    if (pickedDate != null) {
      final TimeOfDay? pickedTime = await showTimePicker(
        context: context,
        initialTime: TimeOfDay.now(),
      );

      if (pickedTime != null) {
        final DateTime fullDateTime = DateTime(
          pickedDate.year,
          pickedDate.month,
          pickedDate.day,
          pickedTime.hour,
          pickedTime.minute,
        );
        _productionTimeController.text =
            DateFormat('yyyy-MM-dd HH:mm').format(fullDateTime);
        widget.setFormData({
          ...widget.formData,
          'production_time': fullDateTime.toIso8601String(),
        });
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Dialog(
      child: Container(
        padding: const EdgeInsets.all(16.0),
        child: Form(
          key: _formKey,
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text(
                widget.modalType == 'create'
                    ? 'Add Raw Milk'
                    : widget.modalType == 'edit'
                        ? 'Edit Raw Milk'
                        : 'Delete Confirmation',
                style: const TextStyle(
                  fontSize: 18,
                  fontWeight: FontWeight.bold,
                ),
              ),
              const SizedBox(height: 16),
              if (widget.isProcessing)
                Column(
                  children: const [
                    CircularProgressIndicator(),
                    SizedBox(height: 8),
                    Text('Processing...'),
                  ],
                )
              else if (widget.modalType == 'delete')
                Text(
                  'Are you sure you want to delete raw milk record "${widget.selectedRawMilk?['cow']['name'] ?? 'this'}"? This action cannot be undone.',
                )
              else
                Column(
                  children: [
                    DropdownButtonFormField<String>(
                      value: widget.formData['cow_id']?.isNotEmpty == true
                          ? widget.formData['cow_id']
                          : null,
                      decoration: const InputDecoration(labelText: 'Cow'),
                      items: widget.cows
                          .where((cow) => cow['gender'] == 'Female')
                          .map<DropdownMenuItem<String>>((cow) {
                        return DropdownMenuItem<String>(
                          value: cow['id'].toString(),
                          child: Text(cow['name']),
                        );
                      }).toList(),
                      onChanged:
                          widget.isProcessing || widget.modalType == 'edit'
                              ? null
                              : (value) {
                                  widget.handleCowChange(value!);
                                },
                      validator: (value) =>
                          value == null ? 'Please select a cow' : null,
                    ),
                    const SizedBox(height: 16),
                    TextFormField(
                      controller: _productionTimeController,
                      decoration: const InputDecoration(
                        labelText: 'Production Time',
                        suffixIcon: Icon(Icons.calendar_today),
                      ),
                      readOnly: true,
                      onTap: () => _selectDateTime(context),
                      validator: (value) => value?.isEmpty ?? true
                          ? 'Please select a time'
                          : null,
                    ),
                    const SizedBox(height: 16),
                    TextFormField(
                      decoration:
                          const InputDecoration(labelText: 'Volume (Liters)'),
                      initialValue:
                          widget.formData['volume_liters']?.toString() ?? '',
                      keyboardType: TextInputType.number,
                      onChanged: (value) {
                        widget.setFormData({
                          ...widget.formData,
                          'volume_liters': double.tryParse(value) ?? 0,
                        });
                      },
                      validator: (value) {
                        if (value?.isEmpty ?? true) {
                          return 'Please enter volume';
                        }
                        if (double.tryParse(value!) == null) {
                          return 'Please enter a valid number';
                        }
                        return null;
                      },
                    ),
                    const SizedBox(height: 16),
                    TextFormField(
                      decoration:
                          const InputDecoration(labelText: 'Previous Volume'),
                      initialValue:
                          widget.formData['previous_volume']?.toString() ?? '0',
                      readOnly: true,
                    ),
                    const SizedBox(height: 16),
                    DropdownButtonFormField<String>(
                      value: widget.formData['lactation_phase'] ?? 'Dry',
                      decoration:
                          const InputDecoration(labelText: 'Lactation Phase'),
                      items: widget.formData['lactation_status'] == true
                          ? const [
                              DropdownMenuItem(
                                value: 'Early',
                                child: Text('Early'),
                              ),
                              DropdownMenuItem(
                                value: 'Mid',
                                child: Text('Mid'),
                              ),
                              DropdownMenuItem(
                                value: 'Late',
                                child: Text('Late'),
                              ),
                            ]
                          : const [
                              DropdownMenuItem(
                                value: 'Dry',
                                child: Text('Dry'),
                              ),
                            ],
                      onChanged: widget.formData['lactation_status'] == true &&
                              !widget.isProcessing
                          ? (value) {
                              widget.setFormData({
                                ...widget.formData,
                                'lactation_phase': value,
                              });
                            }
                          : null,
                    ),
                    Row(
                      children: [
                        Checkbox(
                          value: widget.formData['lactation_status'] ?? false,
                          onChanged: widget.isProcessing
                              ? null
                              : (value) {
                                  widget.setFormData({
                                    ...widget.formData,
                                    'lactation_status': value,
                                    'lactation_phase': value! ? 'Early' : 'Dry',
                                  });
                                },
                        ),
                        const Text('Lactation Status'),
                      ],
                    ),
                  ],
                ),
              const SizedBox(height: 16),
              Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: [
                  TextButton(
                    onPressed: widget.isProcessing
                        ? null
                        : () {
                            widget.setModalType(null);
                          },
                    child: const Text('Cancel'),
                  ),
                  const SizedBox(width: 8),
                  widget.modalType == 'delete'
                      ? ElevatedButton(
                          onPressed:
                              widget.isProcessing ? null : widget.handleDelete,
                          child: const Text('Delete'),
                        )
                      : ElevatedButton(
                          onPressed: widget.isProcessing
                              ? null
                              : () {
                                  if (_formKey.currentState?.validate() ??
                                      false) {
                                    widget.handleSubmit();
                                  }
                                },
                          child: const Text('Save'),
                        ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
